ostd/src/cpu/cpu_local.rs

```diff
@@ -209,13 +223,76 @@ pub(crate) unsafe fn early_init_bsp_local_base() {
 /// It must be guaranteed that the BSP will not access local data before
 /// this function being called, otherwise copying non-constant values
 /// will result in pretty bad undefined behavior.
-pub(crate) unsafe fn init_on_bsp() {
-    // TODO: allocate the pages for application processors and copy the
-    // CPU-local objects to the allocated pages.
+pub unsafe fn init_on_bsp() {
+    let bsp_base_va = __cpu_local_start as usize;
+    let bsp_end_va = __cpu_local_end as usize;
+
+    let nbytes = (bsp_end_va - bsp_base_va).align_up(PAGE_SIZE);
+
+    let num_cpus = super::num_cpus();
+
+    let mut cpu_local_storages = CPU_LOCAL_STORAGES.write();
+    for cpu_i in 1..num_cpus {
+        let ap_pages = page::allocator::alloc_contiguous::<KernelMeta>(nbytes).unwrap();
+        let ap_pages_ptr = paddr_to_vaddr(ap_pages.start_paddr()) as *mut u8;
+
+        // SAFETY: The BSP has not initialized the CPU-local area, so the objects in
+        // in the `.cpu_local` section can be bitwise bulk copied to the AP's local
+        // storage. The destination memory is allocated so it is valid to write to.
+        unsafe {
+            core::ptr::copy_nonoverlapping(
+                bsp_base_va as *const u8,
+                ap_pages_ptr,
+                bsp_end_va - bsp_base_va,
+            );
+        }
+
+        // SAFETY: the first 4 bytes is reserved for storing CPU ID.
+        unsafe {
+            (ap_pages_ptr as *mut u32).write(cpu_i);
+        }
+
+        // SAFETY: the second 4 bytes is reserved for storing the preemt count.
+        unsafe {
+            (ap_pages_ptr as *mut u32).add(1).write(0);
+        }
+
+        cpu_local_storages.push(ap_pages);
+    }
+
+    // Write the CPU ID of BSP to the first 4 bytes of the CPU-local area.
+    let bsp_cpu_id_ptr = bsp_base_va as *mut u32;
+    // SAFETY: the first 4 bytes is reserved for storing CPU ID.
+    unsafe {
+        bsp_cpu_id_ptr.write(0);
+    }
+
+    cpu::local::set_base(bsp_base_va as u64);
 
     #[cfg(debug_assertions)]
-    {
-        IS_INITIALIZED.store(true, Ordering::Relaxed);
+    IS_INITIALIZED.store(true, Ordering::Relaxed);
```

Comment:
I haven't see this variable used.

Reply:
It is a debug assertion performed when getting the address of CPU local data.

https://github.com/asterinas/asterinas/blob/cf1a488132d6e15178e47d9fb5bc632f81edb60c/ostd/src/cpu/cpu_local.rs#L120-L128

Introduced in #1073